src/infinite-tree.js

```diff
@@ -962,10 +962,61 @@ class InfiniteTree extends events.EventEmitter {
 
         this.emit('willOpenNode', node);
 
+        const fn = () => {
+            node.state.open = true; // Set node.state.open to true
+            const openNodes = [node].concat(this.state.openNodes); // the most recently used items first
+            this.state.openNodes = openNodes;
+
+            const nodes = flatten(node.children, { openNodes: this.state.openNodes });
+            const rows = [];
+            // Update rows
+            rows.length = nodes.length;
+            for (let i = 0; i < nodes.length; ++i) {
+                const node = nodes[i];
+                rows[i] = this.options.rowRenderer(node, this.options);
+            }
+
+            // Update nodes & rows
+            this.nodes.splice.apply(this.nodes, [nodeIndex + 1, 0].concat(nodes));
+            this.rows.splice.apply(this.rows, [nodeIndex + 1, 0].concat(rows));
+
+            // Update the row corresponding to the node
+            this.rows[nodeIndex] = this.options.rowRenderer(node, this.options);
+
+            // Add all child nodes to the lookup table if the first child does not exist in the lookup table
+            if ((nodes.length > 0) && !(this.nodeTable.get(nodes[0]))) {
+                nodes.forEach((node) => {
+                    if (node.id !== undefined) {
+                        this.nodeTable.set(node.id, node);
+                    }
+                });
+            }
+
+            // Toggle the expanding state
+            node.state.expanding = false;
+            // Update the row corresponding to the node
+            this.rows[nodeIndex] = this.options.rowRenderer(node, this.options);
+            // Update list
+            this.update();
+
+            if (!silent) {
+                // Emit a "openNode" event
+                this.emit('openNode', node);
+            }
+
+            if (typeof asyncCallback === 'function') {
+                asyncCallback();
+            }
+        };
+
         // Retrieve node index
         const nodeIndex = this.nodes.indexOf(node);
         if (nodeIndex < 0) {
-            error('Invalid node index');
+            if (async) {
```

Comment:
This is not the expected behavior. If the node does not exist, it should not add to `this.state.openNodes`.

Reply:
The point here that node exists but it is collapsed. That's why it is not in *this.nodes*, but it is in *this.treeNode*.
Why we need to do it:
1. Clicking on Lazy-Loaded:
<img width="144" alt="screen shot 2018-01-24 at 07 15 02" src="https://user-images.githubusercontent.com/2265094/35317290-5aecbb1e-00d6-11e8-954a-7d69a178f618.png">
2. Waiting when it loads child, and click on the child 
<img width="184" alt="screen shot 2018-01-24 at 07 15 56" src="https://user-images.githubusercontent.com/2265094/35317301-711a83d0-00d6-11e8-937b-2741846cbf70.png">
3. When it still loading, clicking on collapsing of Lazy-Loaded:
<img width="166" alt="screen shot 2018-01-24 at 07 16 49" src="https://user-images.githubusercontent.com/2265094/35317316-8fd3cc50-00d6-11e8-838c-0cd6bab13444.png">
4. Wait till child of child is loaded, and expanding now Lazy-Loaded. So here i'm expecting such tree:
<img width="215" alt="screen shot 2018-01-24 at 07 16 36" src="https://user-images.githubusercontent.com/2265094/35317333-a638412e-00d6-11e8-88bc-c7bac262d1b7.png">
Why? Because I've already started the process of loading children, why a user needs to reload the whole sub-tree again, if he expanded a child before, but collapsed a node of grand-parent. 
As for me, as a user of a tree, looking again at such tree state would be annoying:
<img width="163" alt="screen shot 2018-01-24 at 07 19 18" src="https://user-images.githubusercontent.com/2265094/35317395-e8edd38a-00d6-11e8-8c0e-2e9d8fc530d2.png">
Don't you agree on that?

Reply:
What I pointed here is that you need to handle non-existing nodes in your PR:

```js
tree.openNode(new Node());
tree.openNode(new Node());
tree.openNode(new Node());
tree.openNode(new Node());

console.log(tree.state.openNodes);
```

which will result in

![image](https://user-images.githubusercontent.com/447801/35319982-188b17c4-011d-11e8-96c7-d59e4e7390c8.png)

Reply:
The nodes in the `tree.state.openNodes` array should be an existed node in the tree. You can check the existence with this.nodeTable.

Reply:
Ok, if that's your point, sure I'll add that check, it makes sense.